var/spack/repos/builtin/packages/hpx/package.py

```diff
@@ -71,61 +72,91 @@ class Hpx(CMakePackage, CudaPackage):
 
-    # Recommended dependency versions for 1.2.X
-    depends_on('cmake@3.9.0:', when='@:1.2.1', type='build')
-    depends_on('boost@1.62.0:', when='@:1.2.1')
-    depends_on('hwloc@1.11:', when='@:1.2.1')
-
-    # Recommended dependency versions before 1.2
-    depends_on('boost@1.55.0:', when='@:1.1.0')
-    depends_on('hwloc@1.6:', when='@:1.1.0')
-
-    # boost 1.73.0 build problem with HPX 1.4.0 and 1.4.1
-    # https://github.com/STEllAR-GROUP/hpx/issues/4728#issuecomment-640685308
-    depends_on('boost@:1.72.0', when='@:1.4')
-
-    # COROUTINES
-    # ~generic_coroutines conflict is not fully implemented
-    # for additional information see:
-    # https://github.com/spack/spack/pull/17654
-    # https://github.com/STEllAR-GROUP/hpx/issues/4829
-    depends_on('boost+context', when='+generic_coroutines')
-    _msg_generic_coroutines = 'This platform requires +generic_coroutines'
-    conflicts('~generic_coroutines', when='platform=darwin', msg=_msg_generic_coroutines)
-
-    # Asio
-    depends_on('asio cxxstd=11', when='@1.7: cxxstd=11')
-    depends_on('asio cxxstd=14', when='@1.7: cxxstd=14')
-    depends_on('asio cxxstd=17', when='@1.7: cxxstd=17')
-
-    # CXX Standard
-    depends_on('boost cxxstd=11', when='cxxstd=11')
-    depends_on('boost cxxstd=14', when='cxxstd=14')
-    depends_on('boost cxxstd=17', when='cxxstd=17')
+    # Other dependecies
+    depends_on('hwloc')
+    depends_on('boost')
+    for cxxstd in cxxstds:
+        depends_on(
+            "boost cxxstd={0}".format(map_cxxstd(cxxstd)),
+            when="cxxstd={0}".format(cxxstd)
+        )
+    depends_on('asio', when='@1.7:')
+    for cxxstd in cxxstds:
+        depends_on(
+            "asio cxxstd={0}".format(map_cxxstd(cxxstd)),
+            when="cxxstd={0} ^asio".format(cxxstd),
+        )
 
-    # Malloc
     depends_on('gperftools', when='malloc=tcmalloc')
     depends_on('jemalloc', when='malloc=jemalloc')
     depends_on('tbb', when='malloc=tbbmalloc')
 
-    # MPI
     depends_on('mpi', when='networking=mpi')
     depends_on('mpi', when='+async_mpi')
 
-    # CUDA
     depends_on('cuda', when='+async_cuda')
 
-    # Instrumentation
     depends_on('otf2', when='instrumentation=apex')
     depends_on('gperftools', when='instrumentation=google_perftools')
     depends_on('papi', when='instrumentation=papi')
     depends_on('valgrind', when='instrumentation=valgrind')
 
+    # Restrictions for stable/master
+    with when("@master"):
+        conflicts("cxxstd=14")
+        depends_on("cuda@11:", when="+cuda")
+
+    with when("@stable"):
+        conflicts("cxxstd=14")
+        depends_on("cuda@11:", when="+cuda")
+
+    # Restrictions for 1.7.X
+    with when('@1.7:'):
+        depends_on('cmake@3.18.0:', type='build')
+        depends_on('boost@1.71.0:')
+        depends_on('asio@1.12.0:')
+        conflicts('%gcc@:6', when='%gcc')
+        conflicts('%clang@:6', when='%clang')
+
+    # Restrictions for 1.6.X
+    conflicts('+rocm', when='@:1.5')
```

Comment:
Just checking, is `rocm` supported in 1.6? or is it a typo?

Reply:
Yep, it's supported in 1.6.0 and not in 1.5.0 and older versions. Ideally I'd like to say that the `+rocm` variant is only available in `@1.6:` but that is not possible as far as I'm aware, hence the backwards conflict.